recipes/lora_finetune_distributed_td.py

```diff
@@ -0,0 +1,1144 @@
+        return load_hf_dataset(
+            **cfg_dataset,
+            transform=transform,
+            streaming=streaming,
+            shuffle=shuffle,
+            parallel_method=parallel_method,
+            num_workers=num_workers,
+        )
+
+    def _setup_data_td(
+        self,
+        cfg_dataloader: DictConfig,
+        cfg_datasets: ListConfig,
+        batch_size: int,
+    ) -> Loader:
+        """
+        All torchdata related setup happens here. Currently this recipe supports
+        both Map and Streaming datasets (from HuggingFace datasets), and mixing multiple
+        datasets (can be mix of Map and Streaming).
+        """
+        # Get global settings
+        shuffle = cfg_dataloader.shuffle
+        parallel_method = cfg_dataloader.get("parallel_method", "thread")
+        packed = cfg_dataloader.get("packed", False)
+        streaming = cfg_dataloader.get("streaming", False)
+        num_workers = cfg_dataloader.get("num_workers", 0)
+        pin_memory = cfg_dataloader.get("pin_memory", True)
+        collate_fn = cfg_dataloader.collate_fn
+        prefetch_factor = cfg_dataloader.get("prefetch_factor", 6)
+
+        # Multi-Dataset Stop Criteria
+        stop_criteria = cfg_dataloader.get(
+            "stop_criteria", StopCriteria.CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED
+        )
+        weights, datasets = {}, {}
+        for idx, cfg_dataset in enumerate(cfg_datasets):
+            dataset_name = cfg_dataset.pop("name", None)
+            if dataset_name is None:
+                dataset_name = cfg_dataset.get("subset", None)
+            key = f"{idx}" + (f"_{dataset_name}" if dataset_name else "")
+            assert key not in weights, f"Duplicate dataset name {key}"
+            weights[key] = float(cfg_dataset.pop("weight"))
+            datasets[key] = self._setup_one_dataset(
+                cfg_dataset=cfg_dataset,
+                global_shuffle=shuffle,
+                global_parallel_method=parallel_method,
+                global_streaming=streaming,
+                global_num_workers=num_workers,
+            )
+
+        # Instantiate collate_fn
+        if "left_pad_sequence" in collate_fn:
+            raise RuntimeError("left_pad_sequence collator is only for inference.")
+
+        collate_fn = (
+            partial(
+                _get_component_from_path(collate_fn),
+                padding_idx=self._tokenizer.pad_id,
+                ignore_idx=self._loss_fn.ignore_index,
+            )
+            if not packed
+            else padded_collate_packed
+        )
+        if len(datasets) > 1:
+            dataset = get_multi_dataset(
+                datasets=datasets,
+                weights=weights,
+                stop_criteria=stop_criteria,
+            )
+        else:
+            dataset = next(iter(datasets.values()))
+
+        loader = get_dataloader(
+            dataset=dataset,
+            model_transform=SFTTransform(model_transform=self._tokenizer),
```

Comment:
What's up with this line? I thought we already pass the transform in `load_hf_dataset`. And why is it just the tokenizer instead of a multimodal transform?